File: web-next/src/components/session/capabilities.ts
import type { ProtocolCapability, ProtocolCapabilitySet } from "@/features/dashboard/types"

export const CAPABILITY = {
  sendMessage: "session.send_message",
  interrupt: "session.interrupt",
  steer: "session.steer",
  approveInteraction: "session.interaction.approval",
  attachment: "runtime.attachment",
  runtimeConfig: "runtime.config",
  modelCatalog: "catalog.model",
  permissionCatalog: "catalog.permission",
  effortCatalog: "catalog.effort",
} as const

export type KnownCapabilityId = (typeof CAPABILITY)[keyof typeof CAPABILITY]

export type RuntimeCapabilityScope = string | {
  runtimeId?: string | null
  runtimeType?: string | null
}

export function findCapability(
  capabilitySet: ProtocolCapabilitySet | null | undefined,
  capabilityId: string,
  runtime?: RuntimeCapabilityScope,
): ProtocolCapability | null {
  const matches = capabilitySet?.capabilities.filter(
    (capability) => capability.capabilityId === capabilityId,
  ) ?? []
  if (!runtime) return matches[0] ?? null
  const runtimeId = typeof runtime === "string" ? runtime : runtime.runtimeId ?? null
  const runtimeType = typeof runtime === "string" ? runtime : runtime.runtimeType ?? null
  return matches.find(
    (capability) => runtimeId && capability.runtimeId === runtimeId,
  )
    ?? matches.find((capability) => runtimeType && capability.runtime === runtimeType)
    ?? matches.find((capability) => !capability.runtime && !capability.runtimeId)
    ?? null
}

export function capabilityIsUsable(
  capabilitySet: ProtocolCapabilitySet | null | undefined,
  capabilityId: KnownCapabilityId,
  runtime?: RuntimeCapabilityScope,
): boolean {
  const capability = findCapability(capabilitySet, capabilityId, runtime)
  return Boolean(capability?.supported && capability.available && capability.allowed)
}

File: server/agent_server/core/capabilities.py
from __future__ import annotations

from typing import Final

from agent_server.core.protocol import ProtocolCapability, ProtocolCapabilitySet

SESSION_SEND_MESSAGE: Final = "session.send_message"
SESSION_INTERRUPT: Final = "session.interrupt"
SESSION_STEER: Final = "session.steer"
SESSION_INTERACTION_APPROVAL: Final = "session.interaction.approval"
RUNTIME_ATTACHMENT: Final = "runtime.attachment"
RUNTIME_CONFIG: Final = "runtime.config"
CATALOG_MODEL: Final = "catalog.model"
CATALOG_PERMISSION: Final = "catalog.permission"
CATALOG_EFFORT: Final = "catalog.effort"


def find_capability(
    capability_set: ProtocolCapabilitySet,
    capability_id: str,
) -> ProtocolCapability | None:
    return next(
        (
            capability
            for capability in capability_set.capabilities
            if capability.capabilityId == capability_id
        ),
        None,
    )


def capability_is_usable(
    capability_set: ProtocolCapabilitySet,
    capability_id: str,
) -> bool:
    capability = find_capability(capability_set, capability_id)
    return bool(
        capability is not None
        and capability.supported
        and capability.available
        and capability.allowed
    )
